colossalai/nn/optimizer/distributed_lamb.py

```diff
@@ -0,0 +1,163 @@
+            tp_group = device_mesh.get_process_group(axis=0)
+            dp_group = device_mesh.get_process_group(axis=1)
+            return device_mesh, tp_group, dp_group
+
+    @torch.no_grad()
+    def step(self, closure=None):
+        """Performs a single optimization step.
+        Arguments:
+            closure (callable, optional): A closure that reevaluates the model
+                and returns the loss.
+        """
+        loss = None
+        if closure is not None:
+            loss = closure()
+
+        torch.nn.utils.clip_grad_norm_(
+            parameters=[p for group in self.param_groups for p in group["params"]], max_norm=1.0, norm_type=2
+        )
+
+        for group in self.param_groups:
+            for p in group["params"]:
+                if p.grad is None:
+                    continue
+                grad = p.grad.data
+                if grad.is_sparse:
+                    raise RuntimeError("Lamb does not support sparse gradients, consider SparseAdam instad.")
+
+                state = self.state[p]
+
+                # State initialization
+                if len(state) == 0:
+                    state["step"] = 0
+                    # Exponential moving average of gradient values
+                    state["exp_avg"] = torch.zeros_like(p.data)
+                    # Exponential moving average of squared gradient values
+                    state["exp_avg_sq"] = torch.zeros_like(p.data)
+
+                exp_avg, exp_avg_sq = state["exp_avg"], state["exp_avg_sq"]
+                beta1, beta2 = group["betas"]
+
+                state["step"] += 1
+
+                # Decay the first and second moment running average coefficient
+                # m_t
+                exp_avg.mul_(beta1).add_(grad, alpha=1 - beta1)
+                # v_t
+                exp_avg_sq.mul_(beta2).addcmul_(grad, grad, value=1 - beta2)
+
+                scaled_lr = group["lr"]
+                if group["bias_correction"]:
+                    bias_correction1 = 1 - beta1 ** state["step"]
+                    bias_correction2 = 1 - beta2 ** state["step"]
+                    # Apply debiasing to lr to avoid broadcast
+                    scaled_lr *= (bias_correction2**0.5) / bias_correction1
+                    # exp_avg.div_(bias_correction1)
+                    # exp_avg_sq.div_(bias_correction2)
+
+                update = exp_avg / exp_avg_sq.sqrt().add(group["eps"])
+                if group["weight_decay"] != 0:
+                    update.add_(p.data, alpha=group["weight_decay"])
+
+                if not self.adam:
+                    w_norm = torch.norm(p)
+                    g_norm = torch.norm(update)
+                    # TODO: Call all-reduce here
+                    if self.device_mesh is not None:
+                        norms = torch.stack([w_norm, g_norm])
+                        if self.dp_group is not None:
+                            dist.all_reduce(norms, group=self.dp_group)
```

Comment:
This is not correct. For L2 norm:
![image](https://github.com/hpcaitech/ColossalAI/assets/23111350/4d8f8295-e39e-4cd6-b7d0-c6f318dc79f0)
We should all-reduce the sum of square of x and then compute square root.